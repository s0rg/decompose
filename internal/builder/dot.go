package builder

import (
	"fmt"
	"hash/fnv"
	"io"
	"strings"

	"github.com/emicklei/dot"

	"github.com/s0rg/decompose/internal/node"
)

const outPort = "out"

// dark28 color scheme from https://www.graphviz.org/doc/info/colors.html
var colors = []string{
	"#1b9e77",
	"#d95f02",
	"#7570b3",
	"#e7298a",
	"#66a61e",
	"#e6ab02",
	"#a6761d",
	"#666666",
}

type DOT struct {
	g        *dot.Graph
	clusters map[string]*dot.Graph
}

func NewDOT() *DOT {
	g := dot.NewGraph(dot.Directed)

	g.Attr("splines", "spline")
	g.Attr("concentrate", dot.Literal("true"))

	return &DOT{
		g:        g,
		clusters: make(map[string]*dot.Graph),
	}
}

func (d *DOT) Name() string {
	return "graphviz-dot"
}

func (d *DOT) AddNode(n *node.Node) error {
	var label, color string

	if n.IsExternal() {
		color = "red"
		label = fmt.Sprintf("external: %s", n.Name)
	} else {
		color = "black"
		label = fmt.Sprintf(
			"%s&#92;nimage: %s&#92;nnet: %s",
			n.Name,
			n.Image,
			strings.Join(n.Networks, ", "),
		)
	}

	if n.Meta != nil {
		if len(n.Meta.Info) > 0 {
			label += "&#92;ninfo: " + n.Meta.Info
		}

		if len(n.Meta.Tags) > 0 {
			label += "&#92;ntags: " + strings.Join(n.Meta.Tags, ",")
		}
	}

	g := d.g

	if n.Cluster != "" {
		sg, ok := d.clusters[n.Cluster]
		if !ok {
			sg = g.Subgraph(n.Cluster, dot.ClusterOption{})
			d.clusters[n.Cluster] = sg
			sg.Node(n.Cluster + "_" + outPort).Label(outPort)
		}

		g = sg
	}

	rb := g.Node(n.ID).Attr(
		"color", color,
	).NewRecordBuilder()

	rb.FieldWithId(label, outPort)
	rb.Nesting(func() {
		for i := 0; i < len(n.Ports); i++ {
			p := n.Ports[i]

			rb.FieldWithId(p.Label(), p.ID())
		}
	})

	_ = rb.Build()

	return nil
}

func (d *DOT) getSrc(id string) (rv dot.Node, out string, ok bool) {
	if rv, ok = d.g.FindNodeById(id); ok {
		return rv, outPort, ok
	}

	sg, ok := d.clusters[id]
	if !ok {
		return
	}

	out = id + "_" + outPort

	if rv, ok = sg.FindNodeById(out); ok {
		return rv, out, ok
	}

	return
}

func (d *DOT) getDst(id string, port *node.Port) (rv dot.Node, out string, ok bool) {
	if rv, ok = d.g.FindNodeById(id); ok {
		return rv, port.ID(), ok
	}

	sg, ok := d.clusters[id]
	if !ok {
		return
	}

	out = id + "_" + port.ID()

	if rv, ok = sg.FindNodeById(out); ok {
		return rv, out, ok
	}

	return sg.Node(out).Label(port.Label()), out, true
}

func (d *DOT) AddEdge(srcID, dstID string, port *node.Port) {
	if srcID == "" || dstID == "" { // fast exit, dot doesnt have default cluster
		return
	}

	src, srcPort, ok := d.getSrc(srcID)
	if !ok {
		return
	}

	dst, dstPort, ok := d.getDst(dstID, port)
	if !ok {
		return
	}

	color := labelColor(port.Label())

	var edge dot.Edge

	if srcPort != outPort {
		edge = d.g.Edge(src, dst, port.Label())
	} else {
		edge = d.g.EdgeWithPorts(src, dst, srcPort, dstPort, port.Label())
	}

	edge.Attr("color", color).Attr("fontcolor", color)
}

func (d *DOT) Write(w io.Writer) error {
	d.g.Write(w)

	return nil
}

func labelColor(label string) (rv string) {
	h := fnv.New64a()

	_, _ = io.WriteString(h, label)

	hash := int(h.Sum64())

	if hash < 0 {
		hash = -hash
	}

	return colors[hash%len(colors)]
}
