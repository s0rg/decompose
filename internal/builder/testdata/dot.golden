digraph  {
	subgraph cluster_s1 {
		label="c1";
		n2[label="out"];
		n13[label="1/tcp"];
		n14[label="2/tcp"];
		n3[color="black",label="<out> 1&#92;nimage: node-image&#92;nnet: test-net&#92;ninfo: info 1&#92;ntags: 1|{<tcp1> 1/tcp|<tcp2> 2/tcp}",shape="record"];
		n4[color="black",label="<out> 2&#92;nimage: node-image&#92;nnet: test-net&#92;ninfo: info 2&#92;ntags: 2|{<tcp1> 1/tcp|<tcp2> 2/tcp}",shape="record"];
		
	}
	subgraph cluster_s8 {
		label="c2";
		n10[color="red",label="<out> external: 2|{<tcp2> 2/tcp}",shape="record"];
		n9[label="out"];
		n11[label="1/tcp"];
		n12[label="2/tcp"];
		
	}
	subgraph cluster_s5 {
		label="c3";
		n6[label="out"];
		n7[color="black",label="<out> 3&#92;nimage: node-image&#92;nnet: test-net&#92;ninfo: info 3&#92;ntags: 3|{<tcp1> 1/tcp|<tcp2> 2/tcp}",shape="record"];
		
	}
	concentrate=true;splines="spline";
	n10:out->n3:tcp1[color="#66a61e",fontcolor="#66a61e",label="1/tcp"];
	n10:out->n3:tcp2[color="#e7298a",fontcolor="#e7298a",label="2/tcp"];
	n10:out->n3:tcp3[color="#7570b3",fontcolor="#7570b3",label="3/tcp"];
	n2->n11[color="#66a61e",fontcolor="#66a61e",label="1/tcp"];
	n2->n12[color="#e7298a",fontcolor="#e7298a",label="2/tcp"];
	n9->n13[color="#66a61e",fontcolor="#66a61e",label="1/tcp"];
	n9->n14[color="#e7298a",fontcolor="#e7298a",label="2/tcp"];
	n3:out->n10:tcp2[color="#e7298a",fontcolor="#e7298a",label="2/tcp"];
	n3:out->n10:tcp3[color="#7570b3",fontcolor="#7570b3",label="3/tcp"];
	n3:out->n4:tcp1[color="#66a61e",fontcolor="#66a61e",label="1/tcp"];
	n3:out->n4:tcp2[color="#e7298a",fontcolor="#e7298a",label="2/tcp"];
	n4:out->n3:tcp1[color="#66a61e",fontcolor="#66a61e",label="1/tcp"];
	n4:out->n3:tcp2[color="#e7298a",fontcolor="#e7298a",label="2/tcp"];
	
}
